substrate-archive/src/wasm_tracing.rs

```diff
@@ -0,0 +1,329 @@
+// Copyright 2017-2021 Parity Technologies (UK) Ltd.
+// This file is part of substrate-archive.
+
+// substrate-archive is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+
+// substrate-archive is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+// GNU General Public License for more details.
+
+// You should have received a copy of the GNU General Public License
+// along with substrate-archive.  If not, see <http://www.gnu.org/licenses/>.
+
+//! The Subscriber implementation for Tracing
+
+use std::collections::HashMap;
+use std::convert::TryInto;
+use std::sync::{
+	atomic::{AtomicU64, Ordering},
+	Arc,
+};
+
+use chrono::{DateTime, Utc};
+use parking_lot::Mutex;
+use serde::{Deserialize, Serialize};
+use tracing::{
+	event::Event,
+	field::{Field, Visit},
+	span::{Attributes, Id, Record},
+	Level, Metadata, Subscriber,
+};
+use tracing_subscriber::CurrentSpan;
+
+use sp_tracing::{WASM_NAME_KEY, WASM_TARGET_KEY, WASM_TRACE_IDENTIFIER};
+use substrate_archive_common::{Result, TracingError};
+
+/// Generic BlockHash type that can be any length.
+type BlockHash = Vec<u8>;
+
+/// The Event a tracing subscriber collects before sending data to the TracingActor.
+#[derive(Debug)]
+pub struct EventMessage {
+	pub block_num: Option<u32>,
+	pub hash: Option<Vec<u8>>,
+	pub name: String,
+	pub target: String,
+	pub level: Level,
+	pub values: TraceData,
+	pub parent_id: Id,
+	pub time: DateTime<Utc>,
+	pub file: Option<String>,
+	pub line: Option<u32>,
+}
+
+/// The message a tracing subscriber collects before sending data to the TracingActor.
+#[derive(Debug, Clone)]
+pub struct SpanMessage {
+	pub id: Id,
+	pub parent_id: Option<Id>,
+	pub name: String,
+	pub target: String,
+	pub level: Level,
+	pub values: TraceData,
+	pub start_time: DateTime<Utc>,
+	pub overall_time: chrono::Duration,
+	pub file: Option<String>,
+	pub line: Option<u32>,
+}
+
+/// Finished Trace Data Format. Ready for insertion into a relational database.
+#[derive(Debug)]
+pub struct Traces {
+	block_num: u32,
+	hash: Vec<u8>,
+	pub spans: Vec<SpanMessage>,
+	pub events: Vec<EventMessage>,
+}
+
+impl Traces {
+	pub fn new(block_num: u32, hash: Vec<u8>, events: Vec<EventMessage>, spans: Vec<SpanMessage>) -> Self {
+		Traces { block_num, hash, spans, events }
+	}
+
+	/// Get the hash these spans come from.
+	pub fn hash(&self) -> Vec<u8> {
+		self.hash.clone()
+	}
+
+	/// Get the block number of the block these spans come from.
+	pub fn block_num(&self) -> u32 {
+		self.block_num
+	}
+}
+
+#[derive(Debug)]
+pub struct SpanEvents {
```

Comment:
small nit and totally a personal preference, but I would call it `SpansAndEvents` otherwise perhaps it seems to imply events belonging to a single span